src/compute/flexbox.rs

```diff
@@ -104,106 +106,139 @@ struct AlgoConstants {
+pub fn compute(
+    tree: &mut impl LayoutTree,
+    node: Node,
+    known_dimensions: Size<Option<f32>>,
+    available_space: Size<AvailableSpace>,
+    run_mode: RunMode,
+    cache_slot: usize,
+) -> Size<f32> {
+    let style = tree.style(node);
+    let has_min_max_sizes = style.min_size.width.is_defined()
         || style.min_size.height.is_defined()
         || style.max_size.width.is_defined()
         || style.max_size.height.is_defined();
 
     // Pull these out earlier to avoid borrowing issues
-    let width_maybe_max = style.min_size.width.maybe_resolve(size.width);
-    let width_maybe_min = style.max_size.width.maybe_resolve(size.width);
-    let height_maybe_max = style.min_size.height.maybe_resolve(size.height);
-    let height_maybe_min = style.max_size.height.maybe_resolve(size.height);
+    let min_size = style.min_size.maybe_resolve(known_dimensions);
+    let max_size = style.max_size.maybe_resolve(known_dimensions);
+    let clamped_style_size = style
+        .size
+        .maybe_resolve(known_dimensions)
+        .zip_map(min_size, |size, min| size.maybe_max(min))
+        .zip_map(max_size, |size, max| size.maybe_min(max));
+
+    if has_min_max_sizes {
+        // NODE_LOGGER.log("FLEX: two-pass");
+        let first_pass = compute_preliminary(
+            tree,
+            node,
+            // style.size.maybe_resolve(known_dimensions),
+            known_dimensions.zip_map(clamped_style_size, |known, style| known.or(style)),
+            available_space,
+            RunMode::ComputeSize,
+            cache_slot == 0, //true,
+        );
 
-    let preliminary_size = if has_root_min_max {
-        let first_pass = compute_preliminary(tree, root, style.size.maybe_resolve(size), size, false, true);
+        let clamped_first_pass_size = first_pass
+            .zip_map(min_size, |size, min| size.maybe_max(min))
+            .zip_map(max_size, |size, max| size.maybe_min(max));
 
         compute_preliminary(
             tree,
-            root,
-            Size {
-                width: first_pass.width.maybe_max(width_maybe_max).maybe_min(width_maybe_min).into(),
-                height: first_pass.height.maybe_max(height_maybe_max).maybe_min(height_maybe_min).into(),
-            },
-            size,
-            true,
-            true,
+            node,
+            known_dimensions.zip_map(clamped_first_pass_size, |known, first_pass| known.or(first_pass.into())),
+            available_space,
+            RunMode::PeformLayout,
+            cache_slot == 0, //true,
         )
     } else {
-        compute_preliminary(tree, root, style.size.maybe_resolve(size), size, true, true)
-    };
+        // NODE_LOGGER.log("FLEX: single-pass");
+        compute_preliminary(
+            tree,
+            node,
+            known_dimensions.zip_map(clamped_style_size, |known, style| known.or(style)),
+            available_space,
+            RunMode::PeformLayout,
+            cache_slot == 0, //true,
```

Comment:
Does cache slot = 0 always mean "primary pass"?

Reply:
If yes, should these indexes be an enum for readability?

Reply:
> Does cache slot index = 0 always mean "primary pass"?

That was the original design on this branch. I mapped primary to 0 and secondary to 1. However, the latest version (the change that caused the huge perf improvement) doesn't actually use the cache slot/index variable at all and instead determines the cache slot based on which parts of known_dimensions are set. So we can get rid of this entirely.